src/sut/dumblog/src/Blocker.hs

```diff
@@ -0,0 +1,39 @@
+module Blocker
+  (Blocker, emptyBlocker, blockUntil, wakeUp)
+where
+
+import Control.Concurrent (threadDelay)
+
+import Control.Concurrent.MVar (MVar, newEmptyMVar, takeMVar, putMVar)
+import Data.IORef
+
+import Data.IntMap (IntMap)
+import qualified Data.IntMap as IntMap
+
+newtype Blocker x = Blocker (IORef (IntMap (MVar x)))
+
+emptyBlocker :: IO (Blocker x)
+emptyBlocker = Blocker <$> newIORef IntMap.empty
+
+blockUntil :: Blocker x -> Int -> IO x
+blockUntil (Blocker b) key = do
+  mv <- newEmptyMVar
+  atomicModifyIORef' b $ \m -> (IntMap.insert key mv m, ())
+  takeMVar mv
+
+wakeUp :: Blocker x -> Int -> x -> IO ()
+wakeUp (Blocker b) key response = do
+  mv <- go 10
+  putMVar mv response
+  where
+    go 0 = error $ "Frontend never added MVar"
```

Comment:
Wonder if, from a library's point of view, it make more sense to return a bool here indicating if `wakeUp` succeeded or not rather than throwing an error?

Reply:
Sure